devcareer/code-jammers-backend
# #174838707 A user should be able to reset their password [WIP]

### What does this PR do?
Allow users to reset their passwords

### Description of the Task to be completed?
- create a recover password route
- create a reset password route
- create a recover password controller
- create a reset password controller
- configure nodemailer to send emails
- configure handbars.js for email template
- Sign up for sendgrid email service
- Write tests for the controllers and routes

### How should this be manually tested?
1. Clone the PR
2. Checkout to the branch ```ft-user-can-reset-password-174838707```
3. Install dependencies using ```npm install```
4. Create a ```.env file``` and define the following:
   * ```DEV_DATABASE_URL = <development database url>```
   * ```TEST_DATABASE_URL = <test database url>```
   * ```DATABASE_URL = <production database url>```
   * ```SENDGRID_API_KEY= <your send grid api key>```
   * ```SENDGRID_EMAIL = <your SENDGRID EMAIL goes here>```
5. Run ```npm run migrate``` to create the database tables
6. Run ```npm run seed``` to seed data to the database
7. Run ```npm run start-dev``` to start the server
8. Go to [Temp Mail](https://temp-mail.org/en/) to get a temporally email for testing
9. Make a POST request to the end point: (http://localhost:3000/api/v1/users/signup) using postman with the following object in the body: ```{ "username": <choose a username>, "email": <the temporal email from temp-mail>, "password": "123456" } ```
10. Make a POST request to the end point: (http://localhost:3000/api/v1/users/recover) with the following object in the body: ```{ "email": <the temporal email from temp-mail> }```
11. Check the inbox on temp-mail for reset password mail
12. Copy the link in the email and use it in postman to make a POST request with the following object in the body: ```{ "newPassword": <your new password>}```

### Link to architecture
[Flowchart](https://docs.google.com/document/d/1dPpdT_tc1y_OmolHHXZf-ghmF554ctTYL01Bg6LTe7k/edit?usp=sharing)

### What are the relevant pivotal tracker stories?

[#174838707](https://www.pivotaltracker.com/story/show/174838707)
